Improving monotone constraints ("Fast" method; linked to #2305, #2717)

CharlesAuguste opened the pull request:
@aldanor @redditur

@guolinke @chivee @jameslamb @huanzhang12 @Laurae2 @StrikerRUS
You were all reviewers of the PRs #2305 and #2717. #2305 was judged to be not merge-able because it is too big. Therefore, in my ultimate comment (https://github.com/microsoft/LightGBM/pull/2305#issuecomment-574560469) I said I would split into smaller PRs easier to merge. This is the second PR in relation with #2305; the first one being #2717 and having been merged.

The goal of this PR is to introduce a new constraining method. This is the "Fast" method from #2305. I introduce a new parameter in the config called "monotone_constraints_method". When this is set to "basic", then the original constraining method is used. When it is set to "intermediate", the new constraining method of this PR is used. (And later, we would like to add an even better constraining method, called "Slow" method, for which this parameter will be set to "advanced").

Here are a few details about the changes I made:
- test_engine.py: improved test to make sure to catch any error and to test both methods;
- Parameters.rst, config.h, config_auto.cpp: added the new config parameter "monotone_constraints_method";
- tree.cpp, tree.h, leaf_splits.hpp: added new utility functions which should be fairly straightforward;
- monotone_constraints.hpp: added the methods to compute the new constraints;
- serial_tree_learner.h, serial_tree_learner.cpp: added code to switch between the new method + a new function to compute best splits, used only with the new fast method.

As I detailed in #2305, this new method yields significantly better results than the existing one, because it is not as constraining. Feel free to ask more details about that. The code in monotone_constraints.hpp may be complicated to understand because the new constraining method is an algorithm that goes through the tree from the split to the root and then from the root to all the leaves, and that updates constraints while trying ignore branches it does need to update to save as much time as possible. I am very happy to give more details or to explain the method further if needed.

I guess this PR is still pretty big, but I hope if is still manageable for you. Please @guolinke let me know if that is okay for you. Thank you very much.

EDIT: Here is the link to the report we made for the original PR. The method we implement in this PR is described in it. https://github.com/microsoft/LightGBM/files/3457826/PR-monotone-constraints-report.pdf

CharlesAuguste commented:
Also there is a single build failing about the docs I added. I am not sure why. Let me know if you have an idea. Thanks

StrikerRUS commented:
@CharlesAuguste 

> Also there is a single build failing about the docs I added. I am not sure why. Let me know if you have an idea. Thanks

Please run `python helpers/parameter_generator.py` to fix the error. I guess it's due to excess one blank line in `Parameters.rst`.

CharlesAuguste commented:
I just did as you said and pushed it to see if it would build. I think there was a build error on your side @StrikerRUS (I see errors like "Exception: Please install CMake first"). Thanks

StrikerRUS commented:
@CharlesAuguste Now docs test is OK. All other builds are failing with the following compilation error:
```
...
[ 84%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/data_parallel_tree_learner.cpp:9:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/parallel_tree_learner.h:15:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/gpu_tree_learner.h:22:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/feature_histogram.hpp:19:
/home/travis/build/microsoft/LightGBM/src/treelearner/monotone_constraints.hpp:119:38: error: 
      no member named 'FeatureMonotone' in 'LightGBM::Dataset'
          learner_state.train_data_->FeatureMonotone(inner_feature);
          ~~~~~~~~~~~~~~~~~~~~~~~~~  ^
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/data_parallel_tree_learner.cpp:9:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/parallel_tree_learner.h:15:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/gpu_tree_learner.h:24:
/home/travis/build/microsoft/LightGBM/src/treelearner/serial_tree_learner.h:37:1: warning: 
      class 'CostEfficientGradientBoosting' was previously declared as a struct
      [-Wmismatched-tags]
class CostEfficientGradientBoosting;
^
/home/travis/build/microsoft/LightGBM/src/treelearner/monotone_constraints.hpp:16:8: note: 
      previous use is here
struct CostEfficientGradientBoosting;
       ^
1 warning and 1 error generated.
CMakeFiles/_lightgbm.dir/build.make:387: recipe for target 'CMakeFiles/_lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o' failed
make[3]: *** [CMakeFiles/_lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/feature_parallel_tree_learner.cpp:8:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/parallel_tree_learner.h:15:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/gpu_tree_learner.h:22:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/feature_histogram.hpp:19:
/home/travis/build/microsoft/LightGBM/src/treelearner/monotone_constraints.hpp:119:38: error: 
      no member named 'FeatureMonotone' in 'LightGBM::Dataset'
          learner_state.train_data_->FeatureMonotone(inner_feature);
          ~~~~~~~~~~~~~~~~~~~~~~~~~  ^
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/feature_parallel_tree_learner.cpp:8:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/parallel_tree_learner.h:15:
In file included from /home/travis/build/microsoft/LightGBM/src/treelearner/gpu_tree_learner.h:24:
/home/travis/build/microsoft/LightGBM/src/treelearner/serial_tree_learner.h:37:1: warning: 
      class 'CostEfficientGradientBoosting' was previously declared as a struct
      [-Wmismatched-tags]
class CostEfficientGradientBoosting;
^
/home/travis/build/microsoft/LightGBM/src/treelearner/monotone_constraints.hpp:16:8: note: 
      previous use is here
struct CostEfficientGradientBoosting;
       ^
1 warning and 1 error generated.
CMakeFiles/_lightgbm.dir/build.make:400: recipe for target 'CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o' failed
make[3]: *** [CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/_lightgbm.dir/all' failed
make[2]: *** [CMakeFiles/_lightgbm.dir/all] Error 2
CMakeFiles/Makefile2:84: recipe for target 'CMakeFiles/_lightgbm.dir/rule' failed
make[1]: *** [CMakeFiles/_lightgbm.dir/rule] Error 2
Makefile:164: recipe for target '_lightgbm' failed
make: *** [_lightgbm] Error 2
```

CharlesAuguste commented:
Oh okay I see. Sorry I was confused because all the other checks were passing when there was an issue with the docs, and the library is building locally for me, so I thought my code was fine. I will try to fix all that. Thanks

StrikerRUS commented:
@CharlesAuguste Sorry about the inconvenience! I guess it have started to fail due to recently merged commits in `master`.

guolinke commented:
FeatureMonotone is removed in master. You can rebase to it.

CharlesAuguste commented:
All checks are passing now @guolinke @StrikerRUS! Thanks for your advice. Let me know what you think of the PR.

guolinke commented:
@CharlesAuguste could I make some changes based on your code, and propose PR in your repo?

CharlesAuguste commented:
Of course yes, I would be very happy with that. Do I need to grant you any access to my repo for that, or is it good? Thanks

guolinke commented:
@CharlesAuguste I can fork your repo and make changes : )

guolinke commented:
I am busy recently, will work on it this weekend.

CharlesAuguste commented:
No problem, thank you very much for spending time on this! :)

StrikerRUS commented on include/LightGBM/config.h:
```diff
@@ -356,6 +356,11 @@ struct Config {
 
+  // alias = monotone_constraining_method
+  // desc = used only if ``monotone_constraints`` is set
+  // desc = monotone constraints method: if set to "basic", the most basic monotone constraints method will be used. It does not slow the library at all, but over-constrains the predictions. If set to "intermediate", a more advanced method will be used, which may slow the library very slightly. However, the intermediate method is much less constraining than the basic method and should significant ...
```
@CharlesAuguste Have you published your paper from the [original comment](https://github.com/microsoft/LightGBM/pull/2305#issue-303471370) anywhere? If not yet, maybe host it at arXiv for now? I think that a link here to it will be very useful for users to understand the difference between methods.

CharlesAuguste replied:
@guolinke @StrikerRUS I am working on uploading the report from the original comment somewhere. Right now we are thinking of arXiv as @StrikerRUS suggested. However, I need to get endorsement for that. I will be working on that but unless you have quick a solution for me, this may take some time. Anyway, I will upload the document and link it on all PR for reference as soon as possible. Thanks

StrikerRUS replied:
@CharlesAuguste Thank you very much! I believe, as a temporary solution (in)direct link to the original file will be just fine.
```
https://github.com/microsoft/LightGBM/files/3457826/PR-monotone-constraints-report.pdf
```

CharlesAuguste replied:
Okay, I will edit my initial comment with the link for now. Thanks

StrikerRUS commented on include/LightGBM/config.h:
```diff
@@ -356,6 +356,11 @@ struct Config {
   // desc = dropout rate: a fraction of previous trees to drop during the dropout
   double drop_rate = 0.1;
 
+  // alias = monotone_constraining_method
```
```suggestion
  // alias = monotone_constraining_method, mc_method
```

guolinke commented on src/treelearner/serial_tree_learner.cpp:
```diff
@@ -765,4 +804,51 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
+
+// this function updates the best split for a leaf
+// it is called only when monotone constraints exist
+void SerialTreeLearner::UpdateBestSplitsFromHistograms(
```
I am confused about this function. What do you want to get in this function?
Why there are two loops over feature_index?
Do you want to update the leaf splits, by different constraints?

CharlesAuguste replied:
Yes exactly, the only goal of this function is to update the best split (called split here), with the new constraints it just got. 

To summarize, first a split is made somewhere. Then the function GoUpToFindLeavesToUpdate updates the constraints of all other splits (if needed). Then the function UpdateBestSplitsFromHistograms updates the best splits in case the constraints associated with the previous best splits were modified. Indeed, if they were, the previous best split may not be the best split anymore.

guolinke commented:
@CharlesAuguste please check my PR in your repo.
I decouple the MC class with tree and tree_learner. 
I feel like the code of `GoUpToFindLeavesToUpdate` is a little bit messy, it would be better if you can make it cleaner.

CharlesAuguste commented:
@guolinke Thank you very much, I will look at your PR. I understand that you find GoUpToFindLeavesToUpdate messy. I will try to think of a way to make it simpler. Thanks a lot!

CharlesAuguste commented:
- I rebased on @guolinke PR on my repo and fixed a few issues;
- I rebased on the current master;
- The build is failing for the R package, but I don't think it it is my fault, it looks like it is linked to #2855;
- @StrikerRUS I added mc_constraints as an alias;
- I still need to think about improving the readability of GoUpToFindLeavesToUpdate. I will try to do that shortly.

CharlesAuguste commented:
@guolinke I tried to refactor GoUpToFindLeavesToUpdate to make it less messy (I created a new function, and changed generic variable names to much more explicit ones). Let me know what you think. Hopefully, the function should much easier to understand now!

Also I formatted the code to standard clang-format because the variables names I used were too long for the lines, but feel free to change the format to anything else you like. Thanks

StrikerRUS commented:
@CharlesAuguste 
> The build is failing for the R package, but I don't think it it is my fault, it looks like it is linked to #2855;

I'm so sorry about that! The fix (#2856) was merged into `master` several minutes ago.

CharlesAuguste commented:
> @CharlesAuguste
> 
> > The build is failing for the R package, but I don't think it it is my fault, it looks like it is linked to #2855;
> 
> I'm so sorry about that! The fix (#2856) was merged into `master` several minutes ago.

Thanks for letting me know! All checks seem good now :)

CharlesAuguste commented:
@guolinke I rebased on master. my PR is a bit messy, sorry about that, let me know if that's an issue (but I guess it is not since I imagine all the commits from this PR will be squashed?). Thanks

CharlesAuguste commented on src/treelearner/serial_tree_learner.cpp:
```diff
@@ -687,4 +697,46 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
+    int real_fidx = train_data_->RealFeatureIndex(feature_index);
+    ComputeBestSplitForFeature(
+        histogram_array_, feature_index, real_fidx,
+        true,
```
@guolinke should I be passing elements of col_sampler_.GetByNode() (that I would call at the start of the function) here instead of true now? That way I could remove the warning about monotone constraints and feature fractions not working together? Thanks

guolinke replied:
Not, the `col_sampler_.GetByNode()` returns the different results in every call.

guolinke commented:
Is this PR ready for review?

aldanor commented:
@guolinke I think it's more or less ready, it incorporates your changes and fixes to your changes, and is rebased on master

guolinke commented on include/LightGBM/tree.h:
```diff
@@ -147,6 +147,31 @@ class Tree {
+
+  inline uint32_t threshold_in_bin(int node_idx) const {
+#ifdef DEBUG
+    CHECK(node_idx >= 0);
```
why need this? could we avoid to pass that index?

CharlesAuguste replied:
We don't need the debug flag, I will remove it.

guolinke commented on src/treelearner/serial_tree_learner.cpp:
```diff
@@ -687,4 +697,46 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
+    if (!col_sampler_.is_feature_used_bytree()[feature_index]) {
+      continue;
+    }
+    if (!histogram_array_[feature_index].is_splittable()) {
```
merge above two `if` into one?

guolinke commented on src/treelearner/serial_tree_learner.cpp:
```diff
@@ -687,4 +697,46 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
+
+  OMP_INIT_EX();
+// find splits
+#pragma omp parallel for schedule(static)
```
```suggestion
#pragma omp parallel for schedule(static) num_threads(share_state_->num_threads)
```

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -26,24 +28,59 @@ struct ConstraintEntry {
+                          bool is_numerical_split, int leaf, int new_leaf,
+                          int8_t monotone_type, double right_output,
+                          double left_output, int, const SplitInfo &,
+                          const std::vector<SplitInfo> &) override {
```
sorry, I use the wrong style config. for the reference and pointers, we should use `T& x` and `T* x`, rather than `T &x` and `T *x`.

guolinke commented on src/treelearner/serial_tree_learner.cpp:
```diff
@@ -716,13 +716,11 @@ void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
-      continue;
-    }
-    if (!histogram_array_[feature_index].is_splittable()) {
+    if (!col_sampler_.is_feature_used_bytree()[feature_index] |
```
typo ? `|`

CharlesAuguste replied:
Yes sorry about that I will correct it now.

CharlesAuguste commented:
Okay @guolinke this is building, and I think I addressed all your comments above. Let me know if I can further improve the PR. Thanks

guolinke commented:
overall looks good to me, expect `GoUpToFindLeavesToUpdate`. I will go through it this week.
BTW, it will be better if:
1. more comments in these `GoXXXToFindLeavesToUpdate`.
2. maybe you can use template arguments to reduce the duplicated codes in these functions. 
like https://github.com/microsoft/LightGBM/blob/6b6709b9b3ae302d4a085aadb327540792a502f0/src/treelearner/feature_histogram.hpp#L780-L786

CharlesAuguste commented:
> overall looks good to me, expect `GoUpToFindLeavesToUpdate`. I will go through it this week.
> BTW, it will be better if:
> 
> 1. more comments in these `GoXXXToFindLeavesToUpdate`.
> 2. maybe you can use template arguments to reduce the duplicated codes in these functions.
>    like https://github.com/microsoft/LightGBM/blob/6b6709b9b3ae302d4a085aadb327540792a502f0/src/treelearner/feature_histogram.hpp#L780-L786

Okay, I will try improving that again shortly.

CharlesAuguste commented:
@guolinke I added some comments to the code, and tried to make it a little bit simpler. I am actually not sure I understand what I should do with templates. What duplicated code would you like me to remove? Otherwise let me know if there is a specific part I can improve.
@StrikerRUS it seems there as an issue with the travis CI build. I am not sure I understand the error, and I am not sure I am the cause as I did not modify the docs since my last commit where the checks were all green. Would you be able to help me figure this out? Thanks

StrikerRUS commented:
@CharlesAuguste Sorry for the inconvenience! `docs` test is rather unstable because it relies on `linkchecker` software package which often produces false positive errors for some reasons.

I re-run that test.

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
+      bool is_split_numerical,
+      const std::vector<int>& features_of_splits_going_up_from_original_leaf,
+      int inner_feature,
+      std::vector<bool>& was_original_leaf_right_child_of_split,
```
use `std::vector<T>*` if you want to change its content inside the function. 
you can use `std::vector<T>& ref_a = *a` in the function, to access the vector.

guolinke replied:
if you don't want to change its content, please add `const`.

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
+      std::vector<uint32_t>& thresholds_of_splits_going_up_from_original_leaf,
+      std::vector<bool>& was_original_leaf_right_child_of_split,
+      int split_feature, const SplitInfo& split_info, uint32_t split_threshold,
+      const std::vector<SplitInfo>& best_split_per_leaf) {
```
the same as above, only use const reference.

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
+  }
+
+  void
+  GoUpToFindLeavesToUpdate(const Tree* tree, int node_idx, int split_feature,
```
the difference between two `GoUpToFindLeavesToUpdate`?

guolinke replied:
if there are similar, could them be merged into one function? if not, maybe different names for them are better.

CharlesAuguste replied:
The functions are doing the same thing, I was just using function overloading to make the call simpler. But I can merge the functions no problem.

CharlesAuguste replied:
I moved that to the Update function, let me know if you are happy with it. Thanks

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +92,368 @@ class LeafConstraints {
   std::vector<ConstraintEntry> entries_;
 };
 
+class FastLeafConstraints : public BasicLeafConstraints {
```
`FastLeafConstraints` is my temporary naming, maybe there is a better name?

CharlesAuguste replied:
maybe IntermediateLeafConstraints as I called the method "intermediate" in the config?

guolinke commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -387,13 +387,13 @@ class FastLeafConstraints : public BasicLeafConstraints {
-    std::vector<bool> was_original_leaf_right_child_of_split;
+    std::vector<int>* features_of_splits_going_up_from_original_leaf = new std::vector<int>();
+    std::vector<uint32_t>* thresholds_of_splits_going_up_from_original_leaf = new std::vector<uint32_t>();
+    std::vector<bool>* was_original_leaf_right_child_of_split = new std::vector<bool>();
```
memory leaking here.  you can use `std::vector<T> a`, and pass `&a` in function call.

CharlesAuguste replied:
That's modified.

CharlesAuguste commented:
@guolinke I think I answered all your comments. Let me know what you think. The library is not building, but I don't think it is an issue with this PR. Thanks

guolinke commented:
Thanks @CharlesAuguste very much! This PR overall looks good to me.
BTW, we are fixing the cpplint issue recently, could you fix them too?
refer to https://travis-ci.org/github/microsoft/LightGBM/jobs/663034557?utm_medium=notification&utm_source=github_status

guolinke commented:
ping @StrikerRUS for additional reviews.

StrikerRUS commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +93,360 @@ class LeafConstraints {
+    }
+#ifdef DEBUG
+    CHECK_GE(new_leaf - 1, 0);
+    CHECK_LT((unsigned int)(new_leaf - 1), node_parent_.size());
```
```suggestion
    CHECK_LT(static_cast<size_t>(new_leaf - 1), node_parent_.size());
```

StrikerRUS commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +93,360 @@ class LeafConstraints {
+      // leaf need to be updated
+      // therefore, for the same feature, there is no use going down from the
+      // second time going up on the right (or on the left)
+      for (unsigned int split_idx = 0;
```
```suggestion
      for (size_t split_idx = 0;
```

StrikerRUS commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +93,360 @@ class LeafConstraints {
+      const std::vector<SplitInfo>& best_split_per_leaf) {
+#ifdef DEBUG
+    CHECK_GE(node_idx, 0);
+    CHECK_LT((unsigned int)node_idx, node_parent_.size());
```
```suggestion
    CHECK_LT(static_cast<size_t>(node_idx), node_parent_.size());
```

StrikerRUS commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +93,360 @@ class LeafConstraints {
+    // the original leaves if so the algorithm should keep going down these nodes
+    // to update constraints
+    if (is_split_numerical) {
+      for (unsigned int i = 0;
```
```suggestion
      for (size_t i = 0;
```

StrikerRUS commented on tests/python_package_test/test_engine.py:
```diff
@@ -995,45 +995,68 @@ def test_init_with_subset(self):
         self.assertEqual(subset_data_4.get_data(), "lgb_train_data.bin")
 
-    def test_monotone_constraint(self):
+    # test if categorical features and monotone features can both be in a dataset without causing issues
```
Is it `TODO` comment?

CharlesAuguste replied:
Yes, I forgot about this one sorry. I am not sure anymore if it is worth having a test for that. I actually think that's fine now, I was more worried about that with my old implementation which was much worse. If you want me to add a python test for that I can do it but otherwise, I think just deleting the comment is fine. Let me know what you think. Thanks

StrikerRUS replied:
I believe that having a simple (not so complicated as the current one, just to be sure it will not crash) test for cat. features + constraints will be better than not having a such test. Maybe it is not so useful for now as you feel confident with the current implementation. But it will help spot issues with possible future refactoring or new constraint methods.

CharlesAuguste replied:
I updated the current test that checks for monotonicity to include a categorical variable. That way if someone wants to improve the handling of categorical variables with monotonic constraints, the test will already be there. Let me know if you are happy with that.

StrikerRUS commented on tests/python_package_test/test_engine.py:
```diff
@@ -995,45 +995,68 @@ def test_init_with_subset(self):
-        }
-        constrained_model = lgb.train(params, trainset)
-        self.assertTrue(is_correctly_constrained(constrained_model))
+        number_of_trials = 10
```
Can we decrease this number to save CI time?

CharlesAuguste replied:
Yes I think setting it to 1 should still easily catch the most obvious errors if that's fine with you.

StrikerRUS replied:
Oh, fine! Then please remove the outer loop for trials.

StrikerRUS commented on tests/python_package_test/test_engine.py:
```diff
@@ -995,45 +995,68 @@ def test_init_with_subset(self):
+                params = {
+                    'min_data': 20,
+                    'num_leaves': 20,
+                    'monotone_constraints': '1,-1,0',
```
```suggestion
                    'monotone_constraints': [1, -1, 0],
```

CharlesAuguste commented:
@StrikerRUS I think I answered all your comments. Let me know if you are happy with the changes. There is 1 check failing, but I am not sure if it is my fault, it looks like it timed out. Can you take a look please? Thanks

StrikerRUS commented on src/treelearner/monotone_constraints.hpp:
```diff
@@ -55,14 +94,360 @@ class LeafConstraints {
+          if (threshold >=
+                  thresholds_of_splits_going_up_from_original_leaf[i] &&
+              !was_original_leaf_right_child_of_split[i]) {
+            keep_going_right = false;
```
Can we speedup here (and in the next `if`) by not running the rest of the loop when the opposite direction is already `false` by something like
```
keep_going_right = false;
if (!keep_going_left) {
  break;
}
```
and will it make any sense?

CharlesAuguste replied:
Yes that make sense, that should speed up the function. I just pushed the change, @StrikerRUS let me know if there is anything else.

StrikerRUS commented:
@CharlesAuguste Thank you for revisiting those loops!

Appveyor failed with rare strange MinGW error (#1818) and I don't have access to Appveyor builds, so I'm going to close-reopen this PR to re-trigger all CI builds.

jameslamb commented:
@CharlesAuguste seconding what others have said. AWESOME contribution! Thanks so much for working through the review process with and for taking the time to contribute.

CharlesAuguste commented:
@guolinke @StrikerRUS @jameslamb Thank you very much for working with me on this!! If you are still okay to keep going, I still have more have more as indicated in my original PR. So if that's okay with you I will open the next PR in the coming days. Next one will be about monotone splits penalization as a way to improve the greediness of the way the trees are built, which does not mix too well with monotone constraints. It should be fairly simple code-wise compared to this PR. Thanks again

StrikerRUS commented:
@CharlesAuguste Sure! Please continue with your plan.
